src/core/google-auth.js

```diff
@@ -72,3 +74,251 @@ XMPPSocialProvider.prototype.login = function(loginOpts, continuation) {
+  2a. If we have a refresh token for that email address, return it.
+  2b. If we don't have a refresh token for that email address, we will have to
+    launch another oauth window which with "approval_prompt=force", so that
+    the user grants us "offline access" again and we can get a refresh token
+ */
+XMPPSocialProvider.prototype.getCredentialsInteractive_ = function() {
+  // First try to get a refresh token via the account chooser
+  return new Promise(function(fulfill, reject) {
+    // Check if there is any refresh token stored.  If not, always force the
+    // approval prompt.  This is a small optimization to ensure that we always
+    // get a refresh token on the 1st login attempt without needing to display
+    // 2 oauth views.
+    this.loadLastRefreshTokenAndEmail_().then(function(data) {
+      var isFirstLoginAttempt = !(data && data.refreshToken);
+      this.tryToGetRefreshToken_(isFirstLoginAttempt, null).then(function(responseObj) {
+        // tryToGetRefreshToken_ should always give us an access_token, even
+        // if no refresh_token is given.
+        var accessToken = responseObj.access_token;
+        if (!accessToken) {
+          reject(new Error('Could not find access_token'));
+        }
+
+        // Get the user's email, needed loading/saving refresh tokens to/from
+        // storage.
+        this.getEmail_(accessToken).then(function(email) {
+          if (responseObj.refresh_token) {
+            // refresh_token was given on first attempt, all done.
+            fulfill({
+                refreshToken: responseObj.refresh_token,
+                accessToken: accessToken,
+                email: email});
+            return;
+          }
+
+          // If no refresh_token is returned, it may mean that the user has already
+          // granted this app a refresh token.  We should first check to see if we
+          // already have a refresh token stored for this user, and if not we should
+          // prompt them again with approval_prompt=force to ensure we get a refresh
+          // token.
+          // Note loadRefreshTokenForEmail_ will fulfill with null if there is
+          // no refresh token saved.
+          this.loadRefreshTokenForEmail_(email).then(function(refreshToken) {
+            if (refreshToken) {
+              // A refresh token had already been saved for this email, done.
+              fulfill({
+                  refreshToken: refreshToken,
+                  accessToken: accessToken,
+                  email: email});
+              return;
+            }
+
+            // No refresh token was returned to us, or has been stored for this
+            // email address (this would happen if the user already granted us
+            // a token but re-installed the chrome app).  Try again forcing
+            // the approval prompt for this email address.
+            this.tryToGetRefreshToken_(true, email).then(function(responseObj) {
```

Comment:
If isFirstLoginAttempt==true, the user will always see the approval prompt for offline access, and we should always get a refresh token unless something goes terribly wrong on the Google server side.  So they shouldn't see the approval prompt twice.  Also the only way that a user will end up getting 2 oauth tabs is if they've already permissioned uProxy, but for some reason we don't have their refresh token stored (e.g. because they re-installed uProxy) - in that case they've already permissioned us, so they won't see the approval prompt on the first oauth attempt with forcePrompt=false (when it's not the first login attempt)